libs/payments/cart/src/lib/checkout.service.ts

```diff
@@ -384,27 +374,135 @@ export class CheckoutService {
+          ? await this.customerManager.retrieve(cart.stripeCustomerId)
+          : undefined;
+        if (!customer) {
+          throw new CartError('Customer not found for cart', { cartId });
+        }
+
+        // Process PayPal payment
+        if (subscription.collection_method === 'send_invoice') {
+          if (!subscription.latest_invoice) {
+            throw new CartError('No latest invoice found for subscription', {
+              subscriptionId: subscription.id,
+            });
+          }
+          try {
+            const latestInvoice = await this.invoiceManager.retrieve(
+              subscription.latest_invoice
+            );
+            await this.invoiceManager.processPayPalInvoice(latestInvoice);
+            await this.postPaySteps(cart, cart.version, subscription, cart.uid);
+          } catch (e) {
+            console.error(e);
+            await this.subscriptionManager.cancel(subscription.id);
+            await this.paypalBillingAgreementManager.cancel(
+              customer.metadata[STRIPE_CUSTOMER_METADATA.PaypalAgreement]
+            );
+            await this.cartManager.finishErrorCart(cartId, {
+              errorReasonId: CartErrorReasonId.Unknown,
+            });
+            return;
+          }
+        } else {
+          // process Stripe payment
+          const paymentIntent =
+            await this.subscriptionManager.getLatestPaymentIntent(subscription);
+          if (!paymentIntent) {
+            throw new CartError('PaymentIntent not found for subscription', {
+              subscriptionId: subscription.id,
+            });
+          }
+          switch (paymentIntent.status) {
+            case 'requires_action':
+              await this.cartManager.setNeedsInputCart(cartId);
+              return;
+            case 'succeeded':
+              if (paymentIntent.payment_method) {
+                await this.customerManager.update(customer.id, {
+                  invoice_settings: {
+                    default_payment_method: paymentIntent.payment_method,
+                  },
+                });
+              } else {
+                throw new CartError(
+                  'Failed to update customer default payment method',
+                  { cartId }
+                );
+              }
+              await this.postPaySteps(
+                cart,
+                cart.version,
+                subscription,
+                cart.uid
+              );
+              return;
+            case 'requires_capture':
+            case 'requires_confirmation':
+            case 'requires_payment_method':
+            case 'processing':
+              msDelay = msDelay * 2;
+              if (cart.state !== CartState.PROCESSING) {
+                await this.cartManager.setProcessingCart(cartId);
+              }
+              break;
+            case 'canceled':
+            default:
+              await this.cartManager.finishErrorCart(cartId, {
```

Comment:
The polling function should throw if an error occurs. It should not be responsible for handling the error.

Note, we have a ticket in M4 to add clean up to the checkout methods. [FXA-10628](https://mozilla-hub.atlassian.net/browse/FXA-10628)

Reply:
no longer relevant, as checkout.service no longer has polling logic